ethcore/src/ethereum/ethash.rs

```diff
@@ -180,14 +183,23 @@ impl Engine for Ethash {
 			let bound_divisor = self.ethash_params.gas_limit_bound_divisor;
 			let lower_limit = gas_limit - gas_limit / bound_divisor + 1.into();
 			let upper_limit = gas_limit + gas_limit / bound_divisor - 1.into();
-			if gas_limit < gas_floor_target {
-				min(gas_floor_target, upper_limit)
+			let gas_limit = if gas_limit < gas_floor_target {
+				let gas_limit = min(gas_floor_target, upper_limit);
+				round_block_gas_limit(gas_limit, lower_limit, upper_limit)
 			} else if gas_limit > gas_ceil_target {
-				max(gas_ceil_target, lower_limit)
+				let gas_limit = max(gas_ceil_target, lower_limit);
+				round_block_gas_limit(gas_limit, lower_limit, upper_limit)
 			} else {
-				max(gas_floor_target, min(min(gas_ceil_target, upper_limit),
-					lower_limit + (header.gas_used().clone() * 6.into() / 5.into()) / bound_divisor))
-			}
+				let total_lower_limit = max(lower_limit, gas_floor_target);
+				let total_upper_limit = min(upper_limit, gas_ceil_target);
+				let gas_limit = max(gas_floor_target, min(total_upper_limit,
```

Comment:
I think it would make more sense to have `max(total_upper_limit, ..` (I know it wasn't like this in previous implementation, but still it would make sense)